src/lib/vasi-sync/src/scchannel.rs

```diff
@@ -0,0 +1,254 @@
+    writer_died: bool,
+    has_sleeper: bool,
+    contents_state: ChannelContentsState,
+}
+
+impl From<u32> for ChannelState {
+    fn from(value: u32) -> Self {
+        let has_sleeper = (value & HAS_SLEEPER) != 0;
+        let writer_died = (value & WRITER_DIED) != 0;
+        let contents_state = ((value & 0xff) as u8).into();
+        ChannelState {
+            has_sleeper,
+            writer_died,
+            contents_state,
+        }
+    }
+}
+
+impl From<ChannelState> for u32 {
+    fn from(value: ChannelState) -> Self {
+        let has_sleeper = if value.has_sleeper { HAS_SLEEPER } else { 0 };
+        let writer_died = if value.writer_died { WRITER_DIED } else { 0 };
+        writer_died | has_sleeper | (value.contents_state as u32)
+    }
+}
+
+struct AtomicChannelState(AtomicU32);
+impl AtomicChannelState {
+    pub fn new() -> Self {
+        Self(AtomicU32::new(
+            ChannelState {
+                has_sleeper: false,
+                writer_died: false,
+                contents_state: ChannelContentsState::Empty,
+            }
+            .into(),
+        ))
+    }
+
+    fn transition_contents_state(
+        &self,
+        from: ChannelContentsState,
+        to: ChannelContentsState,
+        order: sync::atomic::Ordering,
+    ) -> ChannelState {
+        let prev: ChannelState = if to as u32 > from as u32 {
+            let inc = to as u32 - from as u32;
+            self.0.fetch_add(inc, order).into()
+        } else {
+            let dec = from as u32 - to as u32;
+            self.0.fetch_sub(dec, order).into()
+        };
+        assert_eq!(prev.contents_state, from);
+        prev
+    }
+
+    fn set_sleeper(&self, order: sync::atomic::Ordering) -> ChannelState {
+        let prev = ChannelState::from(self.0.fetch_add(HAS_SLEEPER, order));
+        assert!(!prev.has_sleeper);
+        prev
+    }
+
+    fn clear_sleeper(&self, order: sync::atomic::Ordering) -> ChannelState {
+        let prev = ChannelState::from(self.0.fetch_sub(HAS_SLEEPER, order));
+        assert!(prev.has_sleeper);
+        prev
+    }
+
+    fn set_writer_dead(&self, order: sync::atomic::Ordering) -> ChannelState {
+        let prev = ChannelState::from(self.0.fetch_add(WRITER_DIED, order));
+        assert!(!prev.writer_died);
+        prev
+    }
+}
+
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+pub enum SelfContainedChannelError {
+    WriterIsDead,
+}
+
+/// A fairly minimal channel implementation that implements the
+/// [`vasi::VirtualAddressSpaceIndependent`] trait.
+///
+/// Only supports a single message at a time, and panics on unexpected state
+/// transitions (e.g. attempting to send when there is already a pending message
+/// in the channel).
+///
+/// Additionally, no attempt is made to ensure consistency after an operation
+/// that panics.  e.g. if one side of a channel makes an API usage error that
+/// causes a panic, the other side of the channel is likely to panic or
+/// deadlock.
```

Comment:
After thinking through this some more, I think it would've been possible to get UB. I couldn't think of an easy way to avoid, and rather than making all the APIs `unsafe`, I got rid of the `fetch_add` and `fetch_sub` tricks. Resending for another look since it's a substantial change, but I think it's easier to reason about.